server.js

```diff
@@ -258,20 +258,12 @@ module.exports.initExpress = function () {
     logLevel: 'silent',
     logProvider: (_provider) => logger,
     router: async (req) => {
-      try {
-        const match = req.url.match(/^\/pl\/workspace\/([0-9]+)\/container\//);
-        if (!match) throw new Error(`Could not match URL: ${req.url}`);
-        const workspace_id = match[1];
-        const result = await sqldb.queryOneRowAsync(
-          `SELECT hostname FROM workspaces WHERE id = $workspace_id;`,
-          { workspace_id }
-        );
-        const url = `http://${result.rows[0].hostname}/`;
-        return url;
-      } catch (err) {
-        logger.error(`Error in router for url=${req.url}: ${err}`);
-        return 'not-matched';
+      if (!req.workspace_hostname) {
+        throw new Error('Workspace host not found');
       }
+
+      // This value is set in the `selectAndValidateWorkspace` middleware.
+      return `http://${req.workspace_hostname}/`;
```

Comment:
I know we don't like these magic values attached to `req` or `res`, but I felt it was appropriate in this case to avoid duplicate queries: this code will run very frequently and is pretty performance-sensitive.

Reply:
Yep, special cases like this seem fine to me.